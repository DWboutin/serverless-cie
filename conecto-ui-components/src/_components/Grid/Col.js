import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import omit from 'omit.js'

class Col extends React.Component {
  colClasses(col, size) {
    const classes = {}

    classes[`col-${size}-${col.span}`] = col.span
    classes[`col-${size}-offset-${col.offset}`] = col.offset
    classes[`first-${size}`] = col.first
    classes[`last-${size}`] = col.last
    classes[`start-${size}`] = col.start
    classes[`center-${size}`] = col.center
    classes[`end-${size}`] = col.end
    classes[`top-${size}`] = col.top
    classes[`middle-${size}`] = col.middle
    classes[`bottom-${size}`] = col.bottom
    classes[`around-${size}`] = col.around
    classes[`between-${size}`] = col.between
    classes[`hidden-${size}`] = col.hidden

    return classes
  }

  buildClasses() {
    const {
      className,
      fullHeight,
      alignMiddle,
      reverse,
      xs,
      sm,
      md,
      lg,
      xl,
    } = this.props
    const xsProps = typeof xs === 'object' ? this.colClasses(xs, 'xs') : {}
    const smProps = typeof sm === 'object' ? this.colClasses(sm, 'sm') : {}
    const mdProps = typeof md === 'object' ? this.colClasses(md, 'md') : {}
    const lgProps = typeof lg === 'object' ? this.colClasses(lg, 'lg') : {}
    const xlProps = typeof xl === 'object' ? this.colClasses(xl, 'xl') : {}

    const classes = classNames({
      col: true,
      'col-xs': !xs,
      'full-height': fullHeight,
      'align-middle': alignMiddle,
      reverse: reverse,
      ...xsProps,
      ...smProps,
      ...mdProps,
      ...lgProps,
      ...xlProps,
      [`${className}`]: className,
    })

    return classes
  }

  render() {
    const { children } = this.props
    const classes = this.buildClasses()
    const componentProps = omit(this.props, [
      'children',
      'className',
      'fullHeight',
      'alignMiddle',
      'reverse',
      'xs',
      'sm',
      'md',
      'lg',
      'xl',
    ])

    return (
      <div className={classes} {...componentProps}>
        {children}
      </div>
    )
  }
}

const sizePropTypes = PropTypes.shape({
  span: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  offset: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  first: PropTypes.bool,
  last: PropTypes.bool,
  start: PropTypes.bool,
  center: PropTypes.bool,
  end: PropTypes.bool,
  top: PropTypes.bool,
  middle: PropTypes.bool,
  bottom: PropTypes.bool,
  around: PropTypes.bool,
  between: PropTypes.bool,
  hidden: PropTypes.bool,
})

Col.defaultProps = {
  fullHeight: false,
  alignMiddle: false,
  reverse: false,
}

Col.propTypes = {
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  fullHeight: PropTypes.bool,
  alignMiddle: PropTypes.bool,
  reverse: PropTypes.bool,
  xs: sizePropTypes,
  sm: sizePropTypes,
  md: sizePropTypes,
  lg: sizePropTypes,
  xl: sizePropTypes,
}

export default Col
