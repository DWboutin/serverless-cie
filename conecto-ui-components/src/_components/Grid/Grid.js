import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import omit from 'omit.js'

class Grid extends React.Component {
  constructor(props) {
    super(props)

    this.classesPrefix = 'to-grid'
  }

  render() {
    const { children, className, fluid, fullHeight } = this.props
    const classes = classNames({
      [this.classesPrefix]: true,
      container: !fluid,
      'container-fluid': fluid,
      'container-full-height': fullHeight,
      [`${className}`]: className,
    })
    const componentProps = omit(this.props, [
      'children',
      'className',
      'fluid',
      'fullHeight',
    ])

    return (
      <div className={classes} {...componentProps}>
        {children}
      </div>
    )
  }
}

Grid.defaultProps = {
  fluid: false,
  fullHeight: false,
}

Grid.propTypes = {
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  fluid: PropTypes.bool,
  fullHeight: PropTypes.bool,
}

export default Grid
