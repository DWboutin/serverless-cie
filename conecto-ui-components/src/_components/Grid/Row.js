import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import omit from 'omit.js'

class Row extends React.Component {
  constructor(props) {
    super(props)

    this.classesPrefix = 'row'
  }

  rowClasses(row, size) {
    const classes = {}

    classes[`${row.alignement}-${size}`] = row.alignement
    classes[`${row.distribution}-${size}`] = row.distribution

    return classes
  }

  buildClasses() {
    const { className, xs, sm, md, lg, xl, reverse } = this.props
    const xsProps = typeof xs === 'object' ? this.rowClasses(xs, 'xs') : {}
    const smProps = typeof sm === 'object' ? this.rowClasses(sm, 'sm') : {}
    const mdProps = typeof md === 'object' ? this.rowClasses(md, 'md') : {}
    const lgProps = typeof lg === 'object' ? this.rowClasses(lg, 'lg') : {}
    const xlProps = typeof xl === 'object' ? this.rowClasses(xl, 'xl') : {}

    const classes = classNames({
      [this.classesPrefix]: true,
      [`${className}`]: className,
      reverse: reverse,
      ...xsProps,
      ...smProps,
      ...mdProps,
      ...lgProps,
      ...xlProps,
    })

    return classes
  }

  render() {
    const { children } = this.props
    const classes = this.buildClasses()
    const componentProps = omit(this.props, [
      'children',
      'className',
      'reverse',
      'xs',
      'sm',
      'md',
      'lg',
      'xl',
    ])

    return (
      <div className={classes} {...componentProps}>
        {children}
      </div>
    )
  }
}

const sizePropTypes = PropTypes.shape({
  alignement: PropTypes.oneOf([
    'start',
    'center',
    'end',
    'top',
    'middle',
    'bottom',
  ]),
  distribution: PropTypes.oneOf(['around', 'between']),
})

Row.propTypes = {
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  reverse: PropTypes.bool,
  xs: sizePropTypes,
  sm: sizePropTypes,
  md: sizePropTypes,
  lg: sizePropTypes,
  xl: sizePropTypes,
}

export default Row
