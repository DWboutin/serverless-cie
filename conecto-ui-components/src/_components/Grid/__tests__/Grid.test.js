import React from 'react'
import { shallow, mount } from 'enzyme'

import Grid, { Row, Col } from '../../Grid'

global.console = {
  log: console.log,
  error: jest.fn(),
}

describe('Grid', () => {
  it('should match snapshot', () => {
    const wrapper = shallow(<Grid>Test</Grid>)

    expect(wrapper).toMatchSnapshot()
  })

  it('should match snapshot with Row and Col', () => {
    const wrapper = mount(
      <Grid>
        <Row>
          <Col>Test</Col>
        </Row>
      </Grid>
    )

    expect(wrapper).toMatchSnapshot()

    wrapper.unmount()
  })

  it('should have the good prefix class and default', () => {
    const wrapper = shallow(<Grid>Test</Grid>)

    expect(wrapper.instance().classesPrefix).toEqual('to-grid')
    expect(wrapper.hasClass(wrapper.instance().classesPrefix)).toEqual(true)
    expect(wrapper.hasClass('container')).toEqual(true)
  })

  it('should set good classes when fluid', () => {
    const wrapper = shallow(<Grid fluid>Test</Grid>)

    expect(wrapper.hasClass('container')).toEqual(false)
    expect(wrapper.hasClass('container-fluid')).toEqual(true)
  })

  it('should set good classes when fullHeight', () => {
    const wrapper = shallow(<Grid fullHeight>Test</Grid>)

    expect(wrapper.hasClass('container')).toEqual(true)
    expect(wrapper.hasClass('container-full-height')).toEqual(true)
  })

  it('should contain added classes', () => {
    const wrapper = shallow(<Grid className="testClass1 testClass2">Test</Grid>)

    expect(wrapper.hasClass('testClass1')).toEqual(true)
    expect(wrapper.hasClass('testClass2')).toEqual(true)
  })

  it('should contain added styles', () => {
    const wrapper = shallow(<Grid style={{ test: 123 }}>Test</Grid>)

    expect(wrapper.prop('style').test).toEqual(123)
  })

  it('should console.error when missing children', () => {
    shallow(<Grid />)

    expect(global.console.error).toBeCalled()
  })

  it('should console.error when not passing a boolean to "fluid" prop', () => {
    shallow(<Grid fluid="123">Test</Grid>)

    expect(global.console.error).toBeCalled()
  })
})
