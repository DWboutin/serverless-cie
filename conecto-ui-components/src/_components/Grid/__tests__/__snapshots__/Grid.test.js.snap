// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid should match snapshot 1`] = `
<div
  className="to-grid container"
>
  Test
</div>
`;

exports[`Grid should match snapshot with Row and Col 1`] = `
<Grid
  fluid={false}
  fullHeight={false}
>
  <div
    className="to-grid container"
  >
    <Row>
      <div
        className="row"
      >
        <Col
          alignMiddle={false}
          fullHeight={false}
          reverse={false}
        >
          <div
            className="col col-xs"
          >
            Test
          </div>
        </Col>
      </div>
    </Row>
  </div>
</Grid>
`;
